pkg/ring/kv/kv_test.go

```diff
@@ -183,18 +183,26 @@ func TestWatchPrefix(t *testing.T) {
 		<-ch1
 		<-ch2
 
-		// verify that each key was reported once, and keys outside prefix were not reported
+		// verify that each key was reported at most once, and keys outside prefix were not reported
+		// also make sure that we get most notifications, if not all.
+		count := 0
 		for i := 0; i < max; i++ {
 			key := fmt.Sprintf("%s%d", prefix, i)
 
-			if observedKeys[key] != 1 {
+			if observedKeys[key] > 1 {
 				t.Errorf("key %s has incorrect value %d", key, observedKeys[key])
 			}
+			count += observedKeys[key]
 			delete(observedKeys, key)
 		}
 
 		if len(observedKeys) > 0 {
 			t.Errorf("unexpected keys reported: %v", observedKeys)
 		}
+
+		const expectedFactor = 0.75
```

Comment:
I'm not a big fan of this, cause it doesn't guarantee it works as expected (ie. no key has been missed). A trick may be waiting - right before the `for` until the number of observed keys doesn't increase for some time (ie. 500ms). You may do something like this:

```
		// wait until the WatchPrefix() stop gets called, which means all the
		// keys have been observed
		lastObservedKeysCount := len(observedKeys)
		lastObservedKeysChangeAt := time.Now()

		test.Poll(t, 2*time.Second, true, func() interface{} {
			count := len(observedKeys)
			elapsed := time.Now().Sub(lastObservedKeysChangeAt)

			if count == lastObservedKeysCount {
				return elapsed >= 250*time.Millisecond
			}

			lastObservedKeysCount = count
			lastObservedKeysChangeAt = time.Now()
			return false
		})
```

Does it make sense?

Reply:
I've tried to implement your suggestion. Test is strict again (each key once), but more forgiving in terms of timeouts. But if nothing is received in 500ms, we stop.